node/router/src/lib.rs

```diff
@@ -99,17 +98,8 @@ pub struct InnerRouter<N: Network> {
     resolver: RwLock<Resolver>,
     /// The set of trusted peers.
     trusted_peers: HashSet<SocketAddr>,
-    /// The map of connected peer IPs to their peer handlers.
-    connected_peers: RwLock<HashMap<SocketAddr, Peer<N>>>,
-    /// The set of handshaking peers. While `Tcp` already recognizes the connecting IP addresses
-    /// and prevents duplicate outbound connection attempts to the same IP address, it is unable to
-    /// prevent simultaneous "two-way" connections between two peers (i.e. both nodes simultaneously
-    /// attempt to connect to each other). This set is used to prevent this from happening.
-    connecting_peers: Mutex<HashMap<SocketAddr, Option<Peer<N>>>>,
-    /// The set of candidate peer IPs.
-    candidate_peers: RwLock<HashSet<SocketAddr>>,
-    /// The set of restricted peer IPs.
-    restricted_peers: RwLock<HashMap<SocketAddr, Instant>>,
+    /// The collection of both candidate and connected peers.
+    peer_pool: RwLock<HashMap<SocketAddr, Peer<N>>>,
```

Comment:
We write lock this whenever we update `last_seen` for any peer, which happens fairly frequently.  This could cause lock contention on nodes with many peers.

It might make sense to use [`DashMap`](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html) here. Alternatively, we could add a `Mutex` or `RwLock` around every peer entry, so we only have to write-lock the `peer_pool` when adding/removing an entry.

Reply:
Sounds fine. On second thought, it is less invasive to reduce the frequency of calling `update_last_seen`. `Instant` usually has a granularity of 1s, so calling it on every inbound message is a little excessive. 

But that can also be done in another PR.